Recommend poetry in docs (#12475)
matrix-org/synapse: docs/development/dependencies.md

# Managing dependencies with Poetry

> This is a quick cheat sheet for developers on how to use [`poetry`](https://python-poetry.org/).

Author: It might be worth some explanatory/additional background here. E.g.
- broad dependencies versus the locked environment.
- the motivation for locking dependencies at all. 
- a link to the issue where I recommended poetry in the first place
- comparison to other ecosystems (npm/cargo. Ruby? Go?)

Reviewer: TBH I think the only item from this list I might add would be the link to the issue where you recommend poetry-the rest feel a little more like blog-post material vs docs material and may risk overwhelming the user. That being said, this doc is already great and I think people will find it quite helpful.

Author: I've added a "background section" which address some of this. It might be overwhelming but I think it's useful to have examples from the project file and the lockfile. I didn't mention anything in the fourth bullet, and only added a link for the third. Hopefully this is a reasonable compromise? /shrug

`[tool.poetry.dependencies]` section. The rest are transitive dependencies (the
things that our direct dependencies themselves depend on, and so on recursively.)

We maintain a locked list of all our dependencies (transitive included) so that
we can track exactly which version of each dependency appears in a given release.
See [here](https://github.com/matrix-org/synapse/issues/11537#issue-1074469665)
for discussion of why we wanted this for Synapse. We chose to use
[`poetry`](https://python-poetry.org/) to manage this locked list; see
[this comment](https://github.com/matrix-org/synapse/issues/11537#issuecomment-1015975819)
for the reasoning.

The locked dependencies get included in our "self-contained" releases: namely,
our docker images and our debian packages. We also use the locked dependencies
in development and our continuous integration.

Separately, our "broad" dependencies—the version ranges specified in
`pyproject.toml`—are included as metadata in our "sdists" and "wheels" [uploaded
to PyPI](https://pypi.org/project/matrix-synapse). Installing from PyPI or from
the Synapse source tree directly will _not_ use the locked dependencies; instead,

> they'll pull in the latest version of each package available at install time.

Reviewer 2: I'm left wondering why we bother to include it as metadata if it is unused. Is there a way to get it to be used?

Author: The broad dependency ranges are automatically included in the wheel as part of the build process. They _are_ used: it's just that almost all of our dependency ranges are `foo <= 1.2.3`, and pip will default to picking the newest version that satisfies the bounds its given.

Perhaps this is better reworded like this?

```suggestion
Separately, our "broad" dependencies—the version ranges specified in
`pyproject.toml`—are included as metadata in our "sdists" and "wheels" [uploaded
to PyPI](https://pypi.org/project/matrix-synapse). Installing from PyPI or from
the Synapse source tree directly will pull in the latest version of each package which satisfies all version range constraints. This method of installation completely ignores the fixed versions present in the lockfile.
```

Requires-Dist: phonenumbers (>=8.2.0)
```

# Tooling recommendation: direnv

[`direnv`](https://direnv.net/) is a tool for activating environments in your
shell inside a given directory. Its support for poetry is unofficial (a
community wiki recipe only), but works solidly in our experience. We thoroughly
recommend it for daily use. To use it:

1. [Install `direnv`](https://direnv.net/docs/installation.html) - it's likely
   packaged for your system already.
2. Teach direnv about poetry. The [shell config here](https://github.com/direnv/direnv/wiki/Python#poetry)
   needs to be added to `~/.config/direnv/direnvrc` (or more generally `$XDG_CONFIG_HOME/direnv/direnvrc`).
3. Mark the synapse checkout as a poetry project: `echo layout poetry > .envrc`.
4. Convince yourself that you trust this `.envrc` configuration and project.
   Then formally confirm this to `direnv` by running `direnv allow`.

Then whenever you navigate to the synapse checkout, you should be able to run
e.g. `mypy` instead of `poetry run mypy`; `python` instead of
`poetry run python`; and your shell commands will automatically be ran in the
context of poetry's venv, without having to run `poetry shell` beforehand.

> # Rules of thumb:

Reviewer 2: This seems to overlap a lot with the FAQ below. Some of the information is basically duplicated, which I find confusing.

Author: Agreed.  I think it's only the last bullet that isn't included in the FAQ below. I'll move it there and remove the other three bullets.

Author: I've had a go at this in 0649ac01d.

# How do I...

## ...reset my venv to the locked environment?

```shell
poetry install --extras all --remove-untracked
```

## ... inspect the `poetry` virtualenv?

Some suggestions:

```shell
# Current env only
poetry env info
# All envs: this allows you to have e.g. a poetry managed venv for Python 3.7,
# and another for Python 3.10.
poetry env list --full-path
poetry run pip list
```

Note that `poetry show` describes the abstract *lock file* rather than your
on-disk environment. With that said, `poetry show --tree` can sometimes be
useful.

## ...add a new dependency?

Either:
- manually update `pyproject.toml`; then `poetry lock --no-update`; or else
- `poetry add packagename`. See `poetry add --help`; note the `--dev`,
  `--extras` and `--optional` flags in particular.
  - **NB**: this specifies the new package with a version given by a "caret bound". This won't get forced to its lowest version in the old deps CI job: see [this TODO](https://github.com/matrix-org/synapse/blob/4e1374373857f2f7a911a31c50476342d9070681/.ci/scripts/test_old_deps.sh#L35-L39).

Include the updated `pyproject.toml` and `poetry.lock` files in your commit.

## ...remove a dependency?

> I don't think we do this often, and I've not tried it myself, but

Reviewer 2: ```suggestion
This is not done often and is untested, but
```

```shell
poetry remove packagename
```

ought to do the trick. Alternatively, manually update `pyproject.toml` and
`poetry lock --no-update`. Include the updated `pyproject.toml` and poetry.lock`
files in your commit.

## ...update the version range for an existing dependency?

Best done by manually editing `pyproject.toml`, then `poetry lock --no-update`.
Include the updated `pyproject.toml` and `poetry.lock` in your commit.

## ...update a dependency in the locked environment?

Use

```shell
poetry update packagename
```

to use the latest version of `packagename` in the locked environment, without
affecting the broad dependencies listed in the wheel.

> There doesn't seem to be away to do this whilst locking a _specific_ version of

Reviewer 2: ```suggestion
There doesn't seem to be a way to do this whilst locking a _specific_ version of
```
